operator/builtin/input/http/operator.go

```diff
@@ -0,0 +1,216 @@
+package httpevents
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/gorilla/mux"
+	jsoniter "github.com/json-iterator/go"
+	"github.com/observiq/stanza/entry"
+	"github.com/observiq/stanza/operator"
+	"github.com/observiq/stanza/operator/helper"
+)
+
+func init() {
+	operator.Register("http_input", func() operator.Builder { return NewHTTPInputConfig("") })
+}
+
+// HTTPInput is an operator that listens for log entries over http.
+type HTTPInput struct {
+	helper.InputOperator
+
+	server      http.Server
+	json        jsoniter.API
+	maxBodySize int64
+
+	auth authMiddleware
+
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+}
+
+// Start will start listening for log entries over http.
+func (t *HTTPInput) Start() error {
+	ctx, cancel := context.WithCancel(context.Background())
+	t.cancel = cancel
+	t.goListen(ctx)
+	return nil
+}
+
+// Stop will stop listening for log entries over http.
+func (t *HTTPInput) Stop() error {
+	t.cancel()
+	t.wg.Wait()
+	return nil
+}
+
+// goListenn will listen for http connections.
+func (t *HTTPInput) goListen(ctx context.Context) {
+	t.Debugf("using server config: %d", t.server.MaxHeaderBytes)
+
+	t.wg.Add(1)
+
+	entryCreateMethods := []string{"POST", "PUT"}
+
+	m := mux.NewRouter()
+	m.HandleFunc("/", t.goHandleMessages).Methods(entryCreateMethods...)
+
+	if t.auth != nil {
+		t.Debugf("using authentication middleware: %s", t.auth.name())
+		m.Use(t.auth.auth)
+	}
+
+	m.HandleFunc("/health", t.health).Methods("GET")
+
+	t.server.Handler = m
+
+	// shutdown go routine waits for a canceled context before stopping the server
+	go func() {
+		defer t.wg.Done()
+		for {
+			select {
+			case <-ctx.Done():
+				t.Debugf("Triggering http server shutdown")
+				ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+				defer cancel()
+				if err := t.server.Shutdown(ctx); err != nil {
+					t.Errorf("error while shutting down http server: %s", err)
+				}
+				return
+			default:
+				time.Sleep(time.Second * 2)
+			}
+		}
+	}()
+
+	// server go routine runs the http server
+	go func() {
+		t.Debugf("Starting http server on socket %s", t.server.Addr)
+		if err := t.server.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+			t.Errorf("http server failed: %s", err)
+			return
+		}
+		t.Debugf("Http server shutdown finished")
+	}()
+}
+
+// goHandleMessages will handles messages from a http connection by reading the request
+// body and returning http status codes.
+func (t *HTTPInput) goHandleMessages(w http.ResponseWriter, req *http.Request) {
+	ctx, cancel := context.WithCancel(req.Context())
+	t.wg.Add(1)
+	defer t.wg.Done()
```

Comment:
It's unnecessary to use wait groups here. This function is called by the server whenever it handles a request, and therefore it's the server's responsibility to ensure that this exits before the server shuts down. Also, you should never increment a wait group within a goroutine, because nothing guarantees that it will be incremented by the time `wg.wait()` is called outside of that routine. Since the server is handling requests in a separate goroutine, this is technically happening here.